File: frontend/driver-pwa/lib/api/phases.ts
// One endpoint completes every phase: POST /trips/{id}/phases/{phase_event_id}/complete,
// returning the trip's full updated TripDetailResponse rather than just the phase.
import { api } from './client'
import type { Trip } from '@shared/lib/types/trip'
import type { PhaseStatus, PhaseType } from '@shared/lib/types/phase'
import type {
  ConfirmationEvidence, DepartureEvidence,
  LoadingEvidence, PhaseEvidence, UnloadingEvidence,
} from '@/lib/types/evidence-draft'
import type { DriverPosition, LocationWarningAcknowledgement } from '@/lib/types/location'
import type { ActionLocationAssessment, DriverLocationCapture } from '@shared/lib/types/action-location'
import { IS_DEMO_MODE } from '@/lib/constants/env'
import { uploadArtifact, type ArtifactType } from './artifacts'

// idempotency_key lets a resubmitted completion return the current state instead of
// duplicating evidence. driver_captured_at is stamped once at submit time and reused on
// replay, never retaken at flush time — see backend schemas/phases.py.
interface PhaseCompleteRequestBase {
  idempotency_key: string
  driver_captured_at: string
  location_warning_acknowledged_at?: string
  location_warning_reason?: string
}

// Mirrors backend schemas/phases.py's discriminated union. `Extract<PhaseType, '...'>`
// fails compilation if the shared PhaseType union ever drops or renames a member here.
export interface ActivationCompleteRequest extends PhaseCompleteRequestBase {
  phase_type: Extract<PhaseType, 'activation'>
  driver_phone_lat: number
  driver_phone_lng: number
}

// No driver_visual_count: the warehouse scan, not a driver figure, records what was
// loaded — the server derives parcel_count_origin from it.
export interface LoadingCompleteRequest extends PhaseCompleteRequestBase {
  phase_type: Extract<PhaseType, 'loading'>
  // Optional — a paperless warehouse has nothing to photograph, so this must never
  // block completion. Sent explicitly as null rather than omitted.
  linehaul_photo_artifact_id: string | null
}

// No guard_verified_seal/seal_number_confirmed: guards have no accounts. Omitting the
// boolean tells the server no independent confirmation was collected — sending `false`
// would record a CRITICAL seal_mismatch on every trip.
export interface DepartureCompleteRequest extends PhaseCompleteRequestBase {
  phase_type: Extract<PhaseType, 'departure'>
  // Normally null — loading's linehaul photo is now the copy of record. Kept optional so
  // a departure queued offline by an older build can still replay with its photo.
  waybill_photo_artifact_id: string | null
  seal_number: string
  seal_photo_artifact_id: string
}

// Lets submitPhase forward a waybill photo from a pre-existing offline-queue entry
// without giving the live DepartureEvidence draft type dead members.
interface LegacyDepartureWaybillPhoto {
  waybillPhotoDataUrl?: string | null
  waybillPhotoArtifactId?: string | null
}

// Arrival attestation only — no photo, artifact id, or seal; when + where is the record.
export interface InTransitCompleteRequest extends PhaseCompleteRequestBase {
  phase_type: Extract<PhaseType, 'in_transit'>
}

export interface UnloadingCompleteRequest extends PhaseCompleteRequestBase {
  phase_type: Extract<PhaseType, 'unloading'>
  seal_number_at_destination: string
  // Seal as found at destination, intact, before the warehouse breaks it. Required —
  // omitting it 422s. Named for the PhaseEvent column it reuses, not what it depicts.
  gate_photo_artifact_id: string
}

export interface ConfirmationCompleteRequest extends PhaseCompleteRequestBase {
  phase_type: Extract<PhaseType, 'confirmation'>
  pod_photo_artifact_id: string
  pod_signature_artifact_id: string
  // Pallet-grain evidence only — never compared against a parcel count client-side.
  // Null, not 0, when unloading's count was left blank.
  driver_visual_count: number | null
  // No pp_scan_in_count: the server derives it from Parcel.pp_scan_in_at, not the client.
}

// trip_creation is deliberately absent: that row is written before a driver is
// involved, and addressing it 422s server-side by design.
export type PhaseCompleteRequest =
  | ActivationCompleteRequest
  | LoadingCompleteRequest
  | DepartureCompleteRequest
  | InTransitCompleteRequest
  | UnloadingCompleteRequest
  | ConfirmationCompleteRequest

// Must exceed the server's Hedera anchor + DB write/refetch budget (~9s measured), or the
// client aborts a submit the server completes anyway; the offline queue then harmlessly
// retries into the backend's idempotency gate.
const PHASE_SUBMIT_TIMEOUT_MS = 30_000
const LOCATION_PREVIEW_TIMEOUT_MS = 5_000

export type { DriverLocationCapture }

export const previewPhaseLocation = (
  tripId: string,
  phaseEventId: string,
  capture: DriverLocationCapture,
): Promise<ActionLocationAssessment> =>
  api.post<ActionLocationAssessment>(
    `/api/v1/trips/${tripId}/phases/${phaseEventId}/location-preview`,
    {
      driver_phone_lat: capture.lat,
      driver_phone_lng: capture.lng,
      driver_captured_at: capture.captured_at,
      driver_accuracy_metres: capture.accuracy_metres,
    },
    { timeoutMs: LOCATION_PREVIEW_TIMEOUT_MS },
  )

// Always 200 on success, including a recorded mismatch (evidence, not an error) or a
// failed Hedera anchor (fail-open) — callers must read the phase's own status/anchor_status,
// not the HTTP code. 404 trip/phase not found; 409 on a closed trip or out-of-sequence phase.
export const completePhase = (
  tripId: string,
  phaseEventId: string,
  request: PhaseCompleteRequest,
): Promise<Trip> =>
  api.post<Trip>(
    `/api/v1/trips/${tripId}/phases/${phaseEventId}/complete`,
    request,
    { timeoutMs: PHASE_SUBMIT_TIMEOUT_MS },
  )

// Omits the keys rather than sending nulls, so a failed capture can't overwrite a
// position an earlier attempt of this submission already stored.
function driverPosition(position: DriverPosition | null): {
  driver_phone_lat?: number
  driver_phone_lng?: number
  driver_accuracy_metres?: number
} {
  if (position === null) return {}
  return {
    driver_phone_lat: position.lat,
    driver_phone_lng: position.lng,
    ...(position.accuracyM === null ? {} : { driver_accuracy_metres: position.accuracyM }),
  }
}

function locationWarningAcknowledgement(
  acknowledgement: LocationWarningAcknowledgement | null | undefined,
): Pick<PhaseCompleteRequestBase, 'location_warning_acknowledged_at' | 'location_warning_reason'> {
  if (acknowledgement === null || acknowledgement === undefined) return {}
  return {
    location_warning_acknowledged_at: acknowledgement.acknowledgedAt,
    location_warning_reason: acknowledgement.reason,
  }
}

// Uses the artifact id an early upload already produced (lib/hooks/useArtifactUpload.ts),
// or uploads now if that never landed — offline at capture, or a failed request.
async function artifactIdFor(
  tripId: string, artifactType: ArtifactType, readyId: string | null,
  dataUrl: string, capturedAt: string,
): Promise<string> {
  if (readyId !== null) return readyId
  const uploaded = await uploadArtifact({ tripId, artifactType, dataUrl, capturedAt })
  return uploaded.id
}

export interface SubmitPhaseResult {
  ok: boolean
  // Null in demo mode. After the trip's last phase, this is the only response that still
  // carries that phase's receipt id, since /trips/me/active returns null once closed.
  trip: Trip | null
  // The addressed phase's own status after the call — a replay of an already-resolved
  // phase still returns 200, so `ok: true` alone doesn't prove fresh work. Null only if
  // the phase is missing from the returned plan (should not happen).
  phaseStatus: PhaseStatus | null
}

// Demo mode returns a mock success immediately. Otherwise uploads any captured photos as
// evidence artifacts, then calls completePhase() with the resulting artifact ids.
//
// `idempotencyKey` must be generated once per logical submission attempt and reused
// across retries (including offline-queue replay), so the server sees the same key.
// `position` is the driver's silent phone fix at submit time; null is legal for every
// phase except activation, which the backend requires. `driverCapturedAt` is stamped
// once per attempt and reused across retries, so a delayed resend still reports the
// original swipe instant.
export async function submitPhase(
  tripId: string,
  phaseEventId: string,
  phaseType: PhaseType,
  evidence: PhaseEvidence,
  idempotencyKey: string,
  position: DriverPosition | null,
  driverCapturedAt: string,
  acknowledgement: LocationWarningAcknowledgement | null = null,
): Promise<SubmitPhaseResult> {
  if (IS_DEMO_MODE) {
    await new Promise<void>((resolve) => setTimeout(resolve, 400))
    return { ok: true, trip: null, phaseStatus: 'completed' }
  }

  // Every evidence draft carries capturedAt — safe to read without a cast.
  const capturedAt = evidence.capturedAt ?? new Date().toISOString()
  const acknowledgementFields = locationWarningAcknowledgement(acknowledgement)

  let updatedTrip: Trip

  switch (phaseType) {
    case 'activation': {
      // The one phase that cannot proceed without a fix: it records where the trip started.
      if (position === null) {
        throw new Error(
          'Could not get your location. Move to open sky, check that Location is enabled for this app, and swipe again.',
        )
      }
      updatedTrip = await completePhase(tripId, phaseEventId, {
        phase_type: 'activation',
        driver_phone_lat: position.lat,
        driver_phone_lng: position.lng,
        ...(position.accuracyM === null ? {} : { driver_accuracy_metres: position.accuracyM }),
        idempotency_key: idempotencyKey,
        driver_captured_at: driverCapturedAt,
        ...acknowledgementFields,
      })
      break
    }
    case 'loading': {
      // linehaulPhotoId falls back to a submit-time upload if the early upload never landed.
      const e = evidence as LoadingEvidence
      const linehaulPhotoId = e.linehaulPhotoDataUrl !== null
        ? await artifactIdFor(tripId, 'photo', e.linehaulPhotoArtifactId, e.linehaulPhotoDataUrl, capturedAt)
        : null
      updatedTrip = await completePhase(tripId, phaseEventId, {
        phase_type: 'loading',
        ...driverPosition(position),
        linehaul_photo_artifact_id: linehaulPhotoId,
        idempotency_key: idempotencyKey,
        driver_captured_at: driverCapturedAt,
        ...acknowledgementFields,
      })
      break
    }
    case 'departure': {
      const e = evidence as DepartureEvidence
      if (e.sealPhotoDataUrl === null || e.sealNumber === null) {
        throw new Error('Departure evidence incomplete — seal photo and seal number are required.')
      }
      // Not captured any more, but a pre-existing offline entry may still carry one —
      // forwarded when present rather than dropped.
      const legacy = evidence as LegacyDepartureWaybillPhoto
      const legacyWaybillDataUrl = legacy.waybillPhotoDataUrl ?? null
      const [waybillPhotoId, sealPhotoId] = await Promise.all([
        legacyWaybillDataUrl !== null
          ? artifactIdFor(tripId, 'photo', legacy.waybillPhotoArtifactId ?? null, legacyWaybillDataUrl, capturedAt)
          : Promise.resolve(null),
        artifactIdFor(tripId, 'photo', e.sealPhotoArtifactId, e.sealPhotoDataUrl, capturedAt),
      ])
      updatedTrip = await completePhase(tripId, phaseEventId, {
        phase_type: 'departure',
        ...driverPosition(position),
        waybill_photo_artifact_id: waybillPhotoId,
        seal_number: e.sealNumber,
        seal_photo_artifact_id: sealPhotoId,
        idempotency_key: idempotencyKey,
        driver_captured_at: driverCapturedAt,
        ...acknowledgementFields,
      })
      break
    }
    case 'unloading': {
      const e = evidence as UnloadingEvidence
      // Truthiness, not `=== null`: a stale offline entry may have this property absent
      // entirely, and `undefined === null` is false.
      if (e.sealNumberAtDestination === null || !e.sealIntactPhotoDataUrl) {
        throw new Error('Unloading evidence incomplete — seal number and intact seal photo are required.')
      }
      const sealIntactPhotoId = await artifactIdFor(
        tripId, 'photo', e.sealIntactPhotoArtifactId ?? null, e.sealIntactPhotoDataUrl, capturedAt,
      )
      updatedTrip = await completePhase(tripId, phaseEventId, {
        phase_type: 'unloading',
        ...driverPosition(position),
        seal_number_at_destination: e.sealNumberAtDestination,
        gate_photo_artifact_id: sealIntactPhotoId,
        idempotency_key: idempotencyKey,
        driver_captured_at: driverCapturedAt,
        ...acknowledgementFields,
      })
      break
    }
    case 'confirmation': {
      const e = evidence as ConfirmationEvidence
      // Count is optional evidence, not a completion gate — only POD photo + signature
      // required. The signature is already an artifact id: the receiver's own browser
      // rendered the attestation and the server stored it, so there's nothing to upload.
      if (e.podPhotoDataUrl === null || !e.podSignatureArtifactId) {
        throw new Error('Confirmation evidence incomplete — POD photo and receiver confirmation are required.')
      }
      const podPhotoId = await artifactIdFor(
        tripId, 'photo', e.podPhotoArtifactId, e.podPhotoDataUrl, capturedAt,
      )
      const podSignatureId = e.podSignatureArtifactId
      updatedTrip = await completePhase(tripId, phaseEventId, {
        phase_type: 'confirmation',
        ...driverPosition(position),
        pod_photo_artifact_id: podPhotoId,
        pod_signature_artifact_id: podSignatureId,
        // `?? null`: a stale offline entry may have this key missing, not null.
        driver_visual_count: e.driverVisualCount ?? null,
        idempotency_key: idempotencyKey,
        driver_captured_at: driverCapturedAt,
        ...acknowledgementFields,
      })
      break
    }
    case 'in_transit': {
      // No evidence upload — the attestation itself is the submission.
      updatedTrip = await completePhase(tripId, phaseEventId, {
        phase_type: 'in_transit',
        ...driverPosition(position),
        idempotency_key: idempotencyKey,
        driver_captured_at: driverCapturedAt,
        ...acknowledgementFields,
      })
      break
    }
    case 'trip_creation':
      // No request variant server-side; reaching here means a routing bug upstream.
      throw new Error(`submitPhase: "${phaseType}" is never completed by a driver action`)
    default: {
      // Exhaustiveness guard: a new PhaseType fails to compile here.
      const unreachable: never = phaseType
      throw new Error(`submitPhase: unhandled phase type "${String(unreachable)}"`)
    }
  }

  const addressedPhase = updatedTrip.phases.find((p) => p.phase_event_id === phaseEventId) ?? null
  return { ok: true, trip: updatedTrip, phaseStatus: addressedPhase?.status ?? null }
}

File: backend/app/schemas/phases.py
"""Pydantic v2 schemas for PhaseEvent and TrailerGpsSnapshot.

stop_sequence and step_recipe are derived fields, not columns, to match the
frozen PhaseDescriptor contract (parent plan §3.1).
"""

import re
from datetime import datetime
from typing import Annotated, Any, Literal, Optional, Union
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.core.phase_meta import STEP_SLUGS
from app.db.models.enums import AnchorStatus, PhaseStatus, PhaseType

_SEAL_PATTERN = re.compile(r"^[A-Z]{2}-\d{4}$")

# Mirror DB column widths (app/db/models/phases.py) so over-length input is a 422, not a 500.
_IDEMPOTENCY_KEY_MAX_LENGTH = 100
_SEAL_NUMBER_MAX_LENGTH = 100


def _validate_seal_format(v: str) -> str:
    # Normalize before matching so retyped case/whitespace isn't rejected as a bad seal.
    normalized = v.strip().upper()
    if not _SEAL_PATTERN.match(normalized):
        raise ValueError("seal number must be in format XX-#### (e.g. AB-1234)")
    return normalized


class PhaseEventRead(BaseModel):
    """One entry in a trip's committed phase plan, as served to the UI.

    Build via from_event(), not model_validate() — stop_sequence needs a TripStop
    join and step_recipe is derived, neither of which model_validate() can fill in.
    """

    model_config = ConfigDict(from_attributes=True)

    # Wire name is `phase_event_id` (frontend/shared/lib/types/phase.ts contract);
    # TrailerGpsSnapshotBase.phase_event_id below is a separate FK, unaffected by this alias.
    id: UUID = Field(..., serialization_alias="phase_event_id")
    trip_id: UUID
    phase_type: PhaseType
    sequence_number: int
    status: PhaseStatus
    anchor_status: AnchorStatus

    # Null only for trip_creation. in_transit anchors to the stop it departs from.
    trip_stop_id: Optional[UUID] = None
    stop_sequence: Optional[int] = None

    step_recipe: tuple[str, ...] = ()  # capture-component slugs; empty for system-observed phases

    # Non-null while waiting on an external system (today: warehouse scan feed).
    # Derived per request (orchestration/phase_gate.py), never stored.
    blocked_on: Optional[str] = None

    idempotency_key: Optional[str] = None  # driver app's offline-queue entry id, echoed back

    dispatcher_override_user_id: Optional[UUID] = None
    dispatcher_override_note: Optional[str] = None
    driver_phone_lat: Optional[float] = None
    driver_phone_lng: Optional[float] = None
    driver_captured_at: Optional[datetime] = None  # nullable for rows completed pre-Task 0A
    horse_gps_lat: Optional[float] = None
    horse_gps_lng: Optional[float] = None
    pulsit_geofence_confirmed: Optional[bool] = None
    seal_number: Optional[str] = None
    seal_photo_artifact_id: Optional[UUID] = None
    waybill_photo_artifact_id: Optional[UUID] = None
    gate_photo_artifact_id: Optional[UUID] = None
    pod_photo_artifact_id: Optional[UUID] = None
    pod_signature_artifact_id: Optional[UUID] = None
    linehaul_photo_artifact_id: Optional[UUID] = None
    parcel_manifest_snapshot: Optional[Any] = None
    parcel_count_origin: Optional[int] = None
    parcel_count_destination: Optional[int] = None
    driver_visual_count: Optional[int] = None
    event_hash: Optional[str] = None
    blockchain_receipt_id: Optional[UUID] = None
    completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_event(
        cls,
        event: Any,
        *,
        stop_sequence_by_id: dict[UUID, int],
        blocked_on_by_stop: dict[tuple[Any, UUID], Optional[str]] | None = None,
    ) -> "PhaseEventRead":
        """Build a PhaseEventRead from a db.models.phases.PhaseEvent.

        `event` is typed Any to keep schemas free of a db-model import. Every
        production caller must pass blocked_on_by_stop, or blocked_on is silently None.
        """
        read = cls.model_validate(event)
        read.stop_sequence = (
            stop_sequence_by_id.get(event.trip_stop_id)
            if event.trip_stop_id is not None
            else None
        )
        read.step_recipe = STEP_SLUGS[PhaseType(event.phase_type)]
        if blocked_on_by_stop is not None and event.trip_stop_id is not None:
            read.blocked_on = blocked_on_by_stop.get(
                (PhaseType(event.phase_type), event.trip_stop_id)
            )
        return read


class TrailerGpsSnapshotBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    phase_event_id: UUID
    trailer_id: UUID
    pulsit_device_id: str
    lat: float
    lng: float
    captured_at: datetime


class TrailerGpsSnapshotCreate(TrailerGpsSnapshotBase):
    pass


class TrailerGpsSnapshotRead(TrailerGpsSnapshotBase):
    id: UUID
    created_at: datetime


class _PhaseCompleteBase(BaseModel):
    # Resubmitting with the same key returns current state instead of erroring — drivers
    # lose signal, replay is normal. max_length mirrors phase_events.idempotency_key.
    idempotency_key: str = Field(..., min_length=1, max_length=_IDEMPOTENCY_KEY_MAX_LENGTH)
    # Optional: a phase must not fail to record because a fix timed out under a loading-bay
    # roof. ActivationCompleteRequest overrides these as required.
    # POPIA: personal location data, stored in Postgres, never anchored (payload builders
    # in orchestration/phase_service.py are explicit whitelists).
    driver_phone_lat: Optional[float] = Field(default=None, ge=-90, le=90)
    driver_phone_lng: Optional[float] = Field(default=None, ge=-180, le=180)

    # Task 0A: instant the driver's phone submitted, captured client-side at swipe time,
    # independent of when the request reaches the server. Lets corroboration_service tell a
    # live handshake from an offline replay. Optional so pre-0A queued entries still replay.
    driver_captured_at: Optional[datetime] = None

    @field_validator("driver_captured_at")
    @classmethod
    def validate_driver_captured_at_is_timezone_aware(cls, v: Optional[datetime]) -> Optional[datetime]:
        # A naive value would silently be treated as UTC downstream, so reject rather than assume.
        if v is not None and v.tzinfo is None:
            raise ValueError("driver_captured_at must be timezone-aware")
        return v

    @model_validator(mode="after")
    def validate_driver_position_pair(self) -> "_PhaseCompleteBase":
        # Both-or-neither: a lone axis is not a position.
        if (self.driver_phone_lat is None) != (self.driver_phone_lng is None):
            raise ValueError("driver_phone_lat and driver_phone_lng must both be provided or both omitted")
        return self


class ActivationCompleteRequest(_PhaseCompleteBase):
    phase_type: Literal[PhaseType.ACTIVATION]
    # Required here, unlike every other phase: this is the origin-gate arrival position.
    driver_phone_lat: float = Field(..., ge=-90, le=90)
    driver_phone_lng: float = Field(..., ge=-180, le=180)


class LoadingCompleteRequest(_PhaseCompleteBase):
    phase_type: Literal[PhaseType.LOADING]  # D7/T5: the seal is applied at departure, not here
    # Ignored by advance_loading (scan-driven redesign); kept only so offline-queued
    # completions from older clients still replay instead of 422ing forever.
    driver_visual_count: Optional[int] = None
    # Optional: a paperless warehouse has no sheet to hand the driver.
    linehaul_photo_artifact_id: Optional[UUID] = None


class DepartureCompleteRequest(_PhaseCompleteBase):
    phase_type: Literal[PhaseType.DEPARTURE]  # D7/T5: seal is applied here, at exit
    # No longer sent by the driver app (superseded by loading's linehaul_photo_artifact_id);
    # kept so offline-queued completions from older clients still replay.
    waybill_photo_artifact_id: Optional[UUID] = None
    seal_number: str
    seal_photo_artifact_id: UUID
    # Tri-state: None means no independent guard confirmation was collected (the ordinary
    # case, guards have no accounts). Only explicit False means a guard refused/failed to verify.
    guard_verified_seal: Optional[bool] = None
    # Seal number the exit guard re-entered; compared against this request's seal_number.
    # Free-form and unpatterned on purpose — a mistyped confirmation is itself evidence.
    seal_number_confirmed: Optional[str] = Field(default=None, max_length=_SEAL_NUMBER_MAX_LENGTH)

    @field_validator("seal_number")
    @classmethod
    def validate_seal_number(cls, v: str) -> str:
        return _validate_seal_format(v)

    @model_validator(mode="after")
    def validate_distinct_evidence_roles(self) -> "DepartureCompleteRequest":
        if self.waybill_photo_artifact_id == self.seal_photo_artifact_id:
            raise ValueError("waybill and seal evidence must use different artifacts")
        return self


class InTransitCompleteRequest(_PhaseCompleteBase):
    # An attestation ("I have arrived"), not an evidence capture — deliberately no photo
    # or artifact, so it stays outside STEP_SLUGS and ANCHORED_PHASES.
    phase_type: Literal[PhaseType.IN_TRANSIT]


class UnloadingCompleteRequest(_PhaseCompleteBase):
    phase_type: Literal[PhaseType.UNLOADING]
    seal_number_at_destination: str
    # Seal as found at destination, before it's broken — tamper-evidence bookend to
    # departure's seal_photo_artifact_id. Required: it can't be re-photographed once open.
    # Reuses PhaseEvent.gate_photo_artifact_id (previously unused).
    # TODO: driver-pwa currently drops this photo and sends only seal_number_at_destination;
    # its captured photo is also of the seal AFTER breaking, not before — needs reconciling.
    gate_photo_artifact_id: UUID

    @field_validator("seal_number_at_destination")
    @classmethod
    def validate_seal_number(cls, v: str) -> str:
        return _validate_seal_format(v)


class ConfirmationCompleteRequest(_PhaseCompleteBase):
    # Proof of delivery requires both a photo and an on-device signature, not either/or.
    # pp_scan_in_count deliberately removed from the wire (server derives it from
    # Parcel.pp_scan_in_at); the same key in the anchored canonical payload is unchanged
    # and must stay, or hash verification breaks on every historical trip.
    phase_type: Literal[PhaseType.CONFIRMATION]
    pod_photo_artifact_id: UUID
    pod_signature_artifact_id: UUID
    # Captured blind (driver never shown an expected figure) and never reconciled against
    # depot scan counts — the two independent scans already settle that. Optional; a
    # skipped count still anchors as None, keeping verification_service's rebuild reproducible.
    driver_visual_count: Optional[int] = None

    @model_validator(mode="after")
    def validate_distinct_evidence_roles(self) -> "ConfirmationCompleteRequest":
        if self.pod_photo_artifact_id == self.pod_signature_artifact_id:
            raise ValueError("POD photo and signature must use different artifacts")
        return self


# Discriminated union: Pydantic picks the member from `phase_type` and validates it
# properly, so a missing seal_number is still a 422, not a hand-rolled service error.
# trip_creation is deliberately absent — it's written by create_trip before any driver
# is involved, and addressing it gets a 409 from complete_phase()'s dispatch table.
PhaseCompleteRequest = Annotated[
    Union[
        ActivationCompleteRequest,
        LoadingCompleteRequest,
        DepartureCompleteRequest,
        InTransitCompleteRequest,
        UnloadingCompleteRequest,
        ConfirmationCompleteRequest,
    ],
    Field(discriminator="phase_type"),
]
